flink-tests/src/test/java/org/apache/flink/test/checkpointing/RescaleCheckpointManuallyITCase.java

```diff
@@ -322,10 +289,14 @@ public void run(SourceContext<Integer> ctx) throws Exception {
                         counter++;
                     }
                 } else {
-                    if (terminateAfterEmission) {
-                        running = false;
+                    while (completedCheckpointNum.get() <= 1) {
+                        Thread.sleep(50);
+                    }
+                    if (failAfterEmission) {
+                        throw new FlinkRuntimeException(
+                                "Make job fail artificially, to retain completed checkpoint.");
```

Comment:
Why we need to fail the job to get the retained checkpoint?

Reply:
Because `RETAIN_ON_CANCELLATION` only takes effects in the case of  "job is cancelled or fails". 
Replace `MiniCluster.cancel()` with failing the job.